go vet errors

I noticed the following errors from `go vet`: 

```sh
$ go vet ./...
# github.com/jetstack/cert-manager/cmd/webhook/app/testing
cmd/webhook/app/testing/testwebhook.go:106:4: call to (*T).Fatalf from a non-test goroutine
# github.com/jetstack/cert-manager/test/integration/framework
test/integration/framework/helpers.go:68:4: call to (*T).Fatal from a non-test goroutine
# github.com/jetstack/cert-manager/test/integration/certificates
test/integration/certificates/metrics_controller_test.go:77:4: call to (*T).Fatal from a non-test goroutine
# github.com/jetstack/cert-manager/test/integration/webhook
test/integration/webhook/dynamic_authority_test.go:79:4: call to (*T).Fatalf from a non-test goroutine
test/integration/webhook/dynamic_authority_test.go:126:4: call to (*T).Fatalf from a non-test goroutine
test/integration/webhook/dynamic_source_test.go:79:4: call to (*T).Fatalf from a non-test goroutine
test/integration/webhook/dynamic_source_test.go:145:4: call to (*T).Fatalf from a non-test goroutine
# github.com/jetstack/cert-manager/test/integration/ctl
test/integration/ctl/ctl_create_cr_test.go:320:7: call to (*T).Fatal from a non-test goroutine
test/integration/ctl/ctl_create_cr_test.go:328:7: call to (*T).Fatal from a non-test goroutine

```



/kind cleanup

--- a/cmd/webhook/app/testing/testwebhook.go
+++ b/cmd/webhook/app/testing/testwebhook.go
@@ -94,16 +94,16 @@ func StartWebhookServer(t *testing.T, ctx context.Context, args []string) (Serve
 	opts.HealthzPort = 0
 
 	stopCh := make(chan struct{})
-	doneCh := make(chan struct{})
+	errCh := make(chan error)
 	srv, err := app.NewServerWithOptions(log, opts)
 	if err != nil {
 		t.Fatal(err)
 	}
 
 	go func() {
-		defer close(doneCh)
+		defer close(errCh)
 		if err := srv.Run(stopCh); err != nil {
-			t.Fatalf("error running webhook server: %v", err)
+			errCh <- fmt.Errorf("error running webhook server: %v", err)
 		}
 	}()
 
@@ -128,7 +128,10 @@ func StartWebhookServer(t *testing.T, ctx context.Context, args []string) (Serve
 	}
 	return serverOpts, func() {
 		close(stopCh)
-		<-doneCh // Wait for shutdown
+		err := <-errCh // Wait for shutdown
+		if err != nil {
+			t.Fatal(err)
+		}
 		if err := os.RemoveAll(tempDir); err != nil {
 			t.Fatal(err)
 		}

--- a/test/integration/certificates/metrics_controller_test.go
+++ b/test/integration/certificates/metrics_controller_test.go
@@ -70,11 +70,11 @@ func TestMetricsController(t *testing.T) {
 	}
 	server := metricsHandler.NewServer(ln, false)
 
-	doneCh := make(chan struct{})
+	errCh := make(chan error)
 	go func() {
-		defer close(doneCh)
+		defer close(errCh)
 		if err := server.Serve(ln); err != http.ErrServerClosed {
-			t.Fatal(err)
+			errCh <- err
 		}
 	}()
 	defer func() {
@@ -84,7 +84,10 @@ func TestMetricsController(t *testing.T) {
 		if err := server.Shutdown(shutdownCtx); err != nil {
 			t.Fatal(err)
 		}
-		<-doneCh
+		err := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
 	}()
 
 	ctrl, queue, mustSync := controllermetrics.NewController(factory, cmFactory, metricsHandler)

--- a/test/integration/ctl/ctl_create_cr_test.go
+++ b/test/integration/ctl/ctl_create_cr_test.go
@@ -301,14 +301,17 @@ func TestCtlCreateCRSuccessful(t *testing.T) {
 				// by the CLI command yet.
 				// Once it has been created, set the `status.certificate` and `Ready` condition so that the `--fetch-certificate`
 				// part of the command is able to proceed.
-				doneCh := make(chan struct{})
+				errCh := make(chan error)
 				pollCtx, cancelPoll := context.WithCancel(ctx)
 				defer func() {
 					cancelPoll()
-					<-doneCh
+					err := <-errCh
+					if err != nil {
+						t.Fatal(err)
+					}
 				}()
 				go func() {
-					defer close(doneCh)
+					defer close(errCh)
 					err = wait.PollImmediateUntil(time.Second, func() (done bool, err error) {
 						req, err = cmCl.CertmanagerV1().CertificateRequests(test.inputNamespace).Get(pollCtx, test.inputArgs[0], metav1.GetOptions{})
 						if err != nil {
@@ -317,7 +320,8 @@ func TestCtlCreateCRSuccessful(t *testing.T) {
 						return true, nil
 					}, pollCtx.Done())
 					if err != nil {
-						t.Fatal("timeout when waiting for CertificateRequest to be created")
+						errCh <- fmt.Errorf("timeout when waiting for CertificateRequest to be created, error: %v", err)
+						return
 					}
 
 					// CR has been created, try update status
@@ -325,7 +329,7 @@ func TestCtlCreateCRSuccessful(t *testing.T) {
 					req.Status.Certificate = test.crStatus.Certificate
 					req, err = cmCl.CertmanagerV1().CertificateRequests(test.inputNamespace).UpdateStatus(pollCtx, req, metav1.UpdateOptions{})
 					if err != nil {
-						t.Fatal(err)
+						errCh <- err
 					}
 				}()
 			}

--- a/test/integration/framework/helpers.go
+++ b/test/integration/framework/helpers.go
@@ -59,18 +59,21 @@ func NewClients(t *testing.T, config *rest.Config) (kubernetes.Interface, inform
 
 func StartInformersAndController(t *testing.T, factory informers.SharedInformerFactory, cmFactory cminformers.SharedInformerFactory, c controllerpkg.Interface) StopFunc {
 	stopCh := make(chan struct{})
-	doneCh := make(chan struct{})
+	errCh := make(chan error)
 	go func() {
-		defer close(doneCh)
+		defer close(errCh)
 		factory.Start(stopCh)
 		cmFactory.Start(stopCh)
 		if err := c.Run(1, stopCh); err != nil {
-			t.Fatal(err)
+			errCh <- err
 		}
 	}()
 	return func() {
 		close(stopCh)
-		<-doneCh
+		err := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
 	}
 }
 

--- a/test/integration/webhook/dynamic_authority_test.go
+++ b/test/integration/webhook/dynamic_authority_test.go
@@ -67,16 +67,19 @@ func TestDynamicAuthority_Bootstrap(t *testing.T) {
 		Log:             logtesting.TestLogger{T: t},
 	}
 	stopCh := make(chan struct{})
-	doneCh := make(chan struct{})
+	errCh := make(chan error)
 	defer func() {
 		close(stopCh)
-		<-doneCh
+		err := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
 	}()
 	// run the dynamic authority controller in the background
 	go func() {
-		defer close(doneCh)
+		defer close(errCh)
 		if err := auth.Run(stopCh); err != nil && !errors.Is(err, context.Canceled) {
-			t.Fatalf("Unexpected error running authority: %v", err)
+			errCh <- fmt.Errorf("Unexpected error running authority: %v", err)
 		}
 	}()
 
@@ -114,16 +117,19 @@ func TestDynamicAuthority_Recreates(t *testing.T) {
 		Log:             logtesting.TestLogger{T: t},
 	}
 	stopCh := make(chan struct{})
-	doneCh := make(chan struct{})
+	errCh := make(chan error)
 	defer func() {
 		close(stopCh)
-		<-doneCh
+		err := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
 	}()
 	// run the dynamic authority controller in the background
 	go func() {
-		defer close(doneCh)
+		defer close(errCh)
 		if err := auth.Run(stopCh); err != nil && !errors.Is(err, context.Canceled) {
-			t.Fatalf("Unexpected error running authority: %v", err)
+			errCh <- fmt.Errorf("Unexpected error running authority: %v", err)
 		}
 	}()
 

--- a/test/integration/webhook/dynamic_source_test.go
+++ b/test/integration/webhook/dynamic_source_test.go
@@ -20,6 +20,7 @@ import (
 	"context"
 	"crypto/x509"
 	"errors"
+	"fmt"
 	"math/big"
 	"testing"
 	"time"
@@ -67,16 +68,19 @@ func TestDynamicSource_Bootstrap(t *testing.T) {
 		Log: log,
 	}
 	stopCh := make(chan struct{})
-	doneCh := make(chan struct{})
+	errCh := make(chan error)
 	defer func() {
 		close(stopCh)
-		<-doneCh
+		err := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
 	}()
 	// run the dynamic authority controller in the background
 	go func() {
-		defer close(doneCh)
+		defer close(errCh)
 		if err := source.Run(stopCh); err != nil && !errors.Is(err, context.Canceled) {
-			t.Fatalf("Unexpected error running source: %v", err)
+			errCh <- fmt.Errorf("Unexpected error running source: %v", err)
 		}
 	}()
 
@@ -133,16 +137,19 @@ func TestDynamicSource_CARotation(t *testing.T) {
 		Log: log,
 	}
 	stopCh := make(chan struct{})
-	doneCh := make(chan struct{})
+	errCh := make(chan error)
 	defer func() {
 		close(stopCh)
-		<-doneCh
+		err := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
 	}()
 	// run the dynamic authority controller in the background
 	go func() {
-		defer close(doneCh)
+		defer close(errCh)
 		if err := source.Run(stopCh); err != nil && !errors.Is(err, context.Canceled) {
-			t.Fatalf("Unexpected error running source: %v", err)
+			errCh <- fmt.Errorf("Unexpected error running source: %v", err)
 		}
 	}()
 

